docs/proposals/trial-custom-crd.md

```diff
@@ -0,0 +1,189 @@
+
+This approach makes impossible to use other CRDs in trial template, even if they satisfy trial job design.
+The number of various Kubernetes CRDs grows significantly and many users would like to use them in Katib
+(e.g, [katib/issue#1081, support Argo Workflow](https://github.com/kubeflow/katib/issues/1081)).
+Another reason to design unified approach is that CRD controller can have `Go` package versions
+that Katib controller doesn't support (e.g, [katib/issue#1081](https://github.com/kubeflow/katib/issues/1081#issuecomment-635338276)).
+
+That is why we propose a new controller design to support custom CRD in trial job and make Katib usable for various Kubernetes resources.
+To make this possible, we are changing API, trial controller, job provider, mutation webhook, metrics collector.
+
+## Goals
+
+1. Allow dynamic watchers for the custom CRD.
+2. Inject Katib sidecar container on training pod.
+3. Indicate training container for metrics collector execution.
+4. Run metrics collector parser after all pod processes completion.
+5. Get succeeded condition of running CRD.
+6. Verify that `sidecar.istio.io/inject: false` label is added.
+
+## Non-Goals
+
+1. Inject Katib sidecar container on more than one pod simultaneously.
+2. Specify list of succeeded conditions for the custom CRD.
+3. Dynamically add new trial watcher for the custom CRD without Katib restart.
+
+## Implementation
+
+During implementation this feature, we should not brake current Katib controller logic.
+Also, we need to make sure that CI is stable and it does not block other Katib work tasks.
+After completion, we can clean-up redundant code.
+
+### API
+
+To achieve above goals, we introduce these `TrialTemplate` API changes.
+
+```go
+
+// TrialTemplate describes structure of Trial template
+type TrialTemplate struct {
+  // Retain indicates that Trial resources must be not cleanup
+  Retain bool `json:"retain,omitempty"`
+
+  // Source for Trial template (unstructured structure or config map)
+  TrialSource `json:",inline"`
+
+  // List of parameters that are used in Trial template
+  TrialParameters []TrialParameterSpec `json:"trialParameters,omitempty"`
+
+  // Label that determines if pod needs to be injected by Katib sidecar container
+  PrimaryPodLabel map[string]string `json:"primaryPodLabel,omitempty"`
+
+  // Name of training container where training is running
+  PrimaryContainerName string `json:"primaryContainerName,omitempty"`
+
+  // Name of condition when Trial custom resource is succeeded
+  SucceededCondition string `json:"succeededCondition,omitempty"`
+
+}
+```
+
+### Trial controller watchers
+
+In the current design trial controller watches
+[three supported resource](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/trial/trial_controller.go#L94-L125).
+To generate these parameters dynamically when Katib starts, we add additional flag (`-trial-resource`)
```

Comment:
Maybe we can use configMap to define these trial-resources ?

Reply:
Currently, in `katib-config` ConfigMap we can set only Suggestion and Metrics collector settings.
Also, we added Watch only when controller starts: https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/trial/trial_controller.go#L79.
I think later when we implement dynamically wathers update, we can think about better design how we can send these resources to the Controller.